kotlinx-coroutines-core/common/src/channels/BroadcastChannel.kt

```diff
@@ -70,6 +74,333 @@ public fun <E> BroadcastChannel(capacity: Int): BroadcastChannel<E> =
+        // in the registration case, so this `onSend` clause should be re-registered.
+        // The idea is that we keep information that this `onSend` clause is already selected
+        // and finish immediately.
+        @Suppress("UNCHECKED_CAST")
+        element as E
+        // First, check whether this `onSend` clause is already
+        // selected, finishing immediately in this case.
+        lock.withLock {
+            val result = onSendInternalResult.remove(select)
+            if (result != null) { // already selected!
+                // `result` is either `Unit` ot `CHANNEL_CLOSED`.
+                select.selectInRegistrationPhase(result)
+                return
+            }
+        }
+        // Start a new coroutine that performs plain `send(..)`
+        // and tries to select this `onSend` clause at the end.
+        CoroutineScope(select.context).launch(start = CoroutineStart.UNDISPATCHED) {
+            val success: Boolean = try {
+                send(element)
+                // The element has been successfully sent!
+                true
+            } catch (t: Throwable) {
+                // This broadcast is closed :(
+                false
+            }
+            // Mark this `onSend` clause as selected and
+            // try to complete the `select` operation.
+            lock.withLock {
+                // Status of this `onSend` clause should not be presented yet.
+                assert { onSendInternalResult[select] == null }
+                // Success or fail? Put the corresponding result.
+                onSendInternalResult[select] = if (success) Unit else CHANNEL_CLOSED
+                // Try to select this `onSend` clause.
+                select as SelectImplementation<*>
+                val trySelectResult = select.trySelectDetailed(this@BroadcastChannelImpl,  Unit)
+                if (trySelectResult !== TrySelectDetailedResult.REREGISTER) {
+                    // In case of re-registration (this `select` was still
+                    // in the registration phase), the algorithm will invoke
+                    // `registerSelectForSend`. As we stored an information that
+                    // this `onSend` clause is already selected (in `onSendInternalResult`),
+                    // the algorithm, will complete immediately. Otherwise, to avoid memory
+                    // leaks, we must remove this information from the hashmap.
+                    onSendInternalResult.remove(select)
+                }
+            }
+
+        }
+    }
+    private val onSendInternalResult = HashMap<SelectInstance<*>, Any?>() // select -> Unit or CHANNEL_CLOSED
+
+    // ############################
+    // # Closing and Cancellation #
+    // ############################
+
+    override fun close(cause: Throwable?): Boolean = lock.withLock { // protected by lock
+        // Close all subscriptions first.
+        subscribers.forEach { it.close(cause) }
```

Comment:
Here and everywhere: please don't invoke any "resuming" operations (close, cancel, send etc.) under the lock. Dispatchers are free to invoke 'resume' in place. With locks, it's a way to hard-to-trace deadlock.
All our existing implementations use lock only to change the state and only after the lock is released, actual resuming operation takes place

Reply:
Also, please ensure such invariant holds for `ConflatedBufferedChannel`. For now it doesn't seem to be the case

Reply:
Let's ignore this for the obsolete broadcasts. As for `ConflatedBufferedChannel`, I will provide a non-blocking implementation instead.

Reply:
I've added a better non-blocking implementation.